client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java

```diff
@@ -2210,4 +2225,10 @@ public ParsedTimestamp(int year, int month, int day, int hour, int minute, int s
             this.timezone = requireNonNull(timezone, "timezone is null");
         }
     }
+
+    abstract static class CancellableIterator<T>
```

Comment:
nit: `AbstractCanclellablIterator`?

Reply:
But then you probably need an interface also

Reply:
Actually it should be interface. It will be nicer. You ca do sth like

```java
public interface CancellableIterator<T>
        extends Iterator<T>
{
    void cancel();

    static <T> CancellableIterator<T> limit(CancellableIterator<T> iterator, long maxRows)
    {
        class LimitIterator
                extends AbstractIterator<T>
                implements CancellableIterator<T>
        {
            private long currentRow;

            @Override
            public void cancel()
            {
                iterator.cancel();
            }

            @Override
            protected T computeNext()
            {
                if (maxRows > 0 && currentRow >= maxRows) {
                    cancel();
                    return endOfData();
                }
                currentRow++;
                if (iterator.hasNext()) {
                    return iterator.next();
                }
                return endOfData();
            }
        }

        return new LimitIterator();
    }
}

```